include/render.h

```diff
@@ -94,17 +95,75 @@ struct Render_t {
 	bool fullFrame = true;
 };
 
+struct RenderedImage {
+	uint16_t width              = 0;
+	uint16_t height             = 0;
+	bool double_width           = false;
+	bool double_height          = false;
+	bool flip_vertical          = false;
+	Fraction pixel_aspect_ratio = {};
+	uint8_t bits_per_pixel      = 0;
+	uint16_t pitch              = 0;
+	const uint8_t* image_data   = nullptr;
+	const uint8_t* palette_data = nullptr;
+
+	inline bool is_paletted() const
+	{
+		return (bits_per_pixel == 8);
+	}
+
+	RenderedImage deep_copy() const
+	{
+		RenderedImage copy = *this;
+
+		// Deep-copy image and palette data
+		const auto image_data_num_bytes = height * pitch;
+
+		copy.image_data = static_cast<uint8_t*>(
+		        std::malloc(image_data_num_bytes));
+		assert(copy.image_data);
+
+		assert(image_data);
+		std::memcpy(const_cast<uint8_t*>(copy.image_data),
+		            image_data,
+		            image_data_num_bytes);
+
+		// TODO it's bad that we need to make this assumption downstream
+		// on the size and alignment of the palette...
+		if (palette_data) {
+			constexpr auto PaletteNumBytes = 256 * 4;
+
+			copy.palette_data = static_cast<uint8_t*>(
+			        std::malloc(PaletteNumBytes));
+			assert(copy.palette_data);
+
+			std::memcpy(const_cast<uint8_t*>(copy.palette_data),
+			            palette_data,
+			            PaletteNumBytes);
+		}
+		return copy;
+	}
+
+	void free()
+	{
+		std::free(const_cast<uint8_t*>(image_data));
+		std::free(const_cast<uint8_t*>(palette_data));
+	}
+};
+
```

Comment:
Prefer automatic memory management e.g. `std::vector<uint8_t>` to manual `malloc` and `free`?

Reply:
This is just a wrapper around existing DOSBox render internals. Plus the deep-copy stuff because we need to pass copies of it to the capture worker threads. Not gonna rewrite the VGA render code now to use vector, sorry...